Absorb Reflect

csis1945 opened the pull request:
You should probably test this before merging.

It works but the timescaling and the positions the boxes spawn in could use some adjustment

look in absorb_reflect scene
projectile shoots at center eye anchor at a collider with tag = "Player"
boxes spawn to show options
hit box with hand to choose
right now absorb destroys the projectile and
reflect does nothing because it reflects by default
time scaling and position of boxes will have to be adjusted
There might be a problem with this commit because 672 changes seems like
a lot. Could you test this scene?
Also I added a heal spell, spell costs don't seem to be implemented yet?

Commit:
Slow/Absorb shows up as boxes that you can hit instead of being part of the gui

Commit:
Absorb/Reflect working with OVR

look in absorb_reflect scene
projectile shoots at center eye anchor at a collider with tag = "Player"
boxes spawn to show options
hit box with hand to choose
right now absorb destroys the projectile and
reflect does nothing because it reflects by default
time scaling and position of boxes will have to be adjusted
There might be a problem with this commit because 672 changes seems like
a lot. Could you test this scene?
Also I added a heal spell, spell costs don't seem to be implemented yet?

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -129,8 +133,16 @@ void updateSigilAnchor()
-
+        currentProjectile = other.gameObject;
         if (getSpeed() > speedThreshold) {
+            if (!this.GetComponent<Hand>().IS_PRIMARY)
```
This needs to stop. You are calling me.findMe.doThing.

"this" IS THIS SCRIPT.
the "hand" script of "this" IS ALSO THIS SCRIPT!

IS_PRIMARY is an IN-SCOPE LOCAL VARIABLE. You are calling a local variable through two layers of abstraction!

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -129,8 +133,16 @@ void updateSigilAnchor()
+            {
+                if (other.gameObject.tag == "Projectile")
+                {
+                    this.GetComponent<Hand>().counterProjectile();
```
NO, please

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -129,8 +133,16 @@ void updateSigilAnchor()
+                if (other.gameObject.tag == "Projectile")
+                {
+                    this.GetComponent<Hand>().counterProjectile();
+                    other.gameObject.GetComponent<Projectile>().reflect();
```
you don't need the gameobject reference, GetComponent is in scope on all monobehaviours, as well as all components.

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -164,4 +176,33 @@ public void switchPrimaryHand()
+        Time.fixedDeltaTime = 0.02f * Time.timeScale;
+        if (GameObject.FindGameObjectWithTag("slow") == null) {
+            GameObject x = Instantiate<GameObject>(Resources.Load<GameObject>("Prefabs/CounterProjectile"));
+            x.transform.position = GameObject.FindGameObjectWithTag("Player").transform.forward*1.1f;
```
just use their hand, which is this script is attached to. you can spawn it right there without any effort.

linxcat commented:
There are too many Finds and Gets in your code. If you are going to be consistently referenceing something, don't call an engine-wide search every single time you need to get a transform. Save it, use it repeatedly.

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -164,4 +176,33 @@ public void switchPrimaryHand()
+        timeSlowed -= Time.deltaTime;
+        if (timeSlowed < 0) {
+            Time.timeScale = 1f;
+            Destroy(GameObject.FindGameObjectWithTag("slow"));
```
this is INCREDIBLY unsafe! You're deleting whatever you find first. You should be calling a function on a target object.

linxcat reviewed (commented):
Lots of cruft in this code. Needs rework

Commit:
Made suggested changes

Should still work the same, I'll test it when I get into 386 today

linxcat commented on Assets/Scripts/Characters/EnemyRanged.cs:
```diff
@@ -130,6 +130,7 @@ enum rangedState {
         GameObject x = (GameObject)Instantiate(projectile);
         x.transform.position = transform.position + new Vector3(0, 1f, 0);
         x.GetComponent<Projectile>().setTarget(player);
```
get it once, use it twice.

linxcat commented on Assets/Scripts/Characters/Projectile.cs:
```diff
@@ -5,25 +5,39 @@
+    // Use this for initialization
+    void Start () {
+
+        target = GameObject.FindGameObjectWithTag("Player");
```
you have a player reference, why aren't you using it?

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -10,6 +10,9 @@ public class Hand : MonoBehaviour {
     float SIGIL_DISTANCE = 0.05F;
     float HIT_ARRAY_DISTANCE = 0.6F;
     float HIT_ARRAY_VERT_BIAS = 0.2F;
+    static GameObject currentProjectile = null;
```
This static means we can only ever have one projectile. I don't think that is intended behaviour

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -129,8 +134,16 @@ void updateSigilAnchor()
+                if (other.tag == "Projectile")
+                {
+                    counterProjectile();
+                    GetComponent<Projectile>().reflect();
```
you are calling this in the current scope, you should be calling other.GetComponent if the function is inside the projectile. The hand does not have a Projectile Component to retreive, as far as I know.

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -164,4 +177,35 @@ public void switchPrimaryHand()
+        }
+    }
+
+    void updateCounterSlowTime() {
```
you could use a coroutine rather than calling this in update every frame. Another, possibly better option is the Invoke command, which calls a functino after n seconds.

linxcat reviewed (changes requested):
Couple more changes, some errors, some design decisions

Commit:
Made changes

Projectile is no longer homing
Also I was not sure what you meant by  the get it once, use it twice
comment.
Making the code in update be a coroutine broke my code so I didn't do
it.

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -136,15 +145,18 @@ void updateSigilAnchor()
         currentProjectile = other.gameObject;
         if (getSpeed() > speedThreshold) {
-            if (!IS_PRIMARY)
+            if (!GetComponent<Hand>().IS_PRIMARY)
```
It was fine before, you changed it to a worse version. This IS the Hand script, you don't need to get it.

linxcat commented on Assets/Scripts/Interaction/Hand.cs:
```diff
@@ -136,15 +145,18 @@ void updateSigilAnchor()
                 {
-                    counterProjectile();
-                    GetComponent<Projectile>().reflect();
+                    GetComponent<Hand>().counterProjectile();
```
why did you add this? It wasn't a problem!

linxcat commented on Assets/Scripts/Characters/Projectile.cs:
```diff
@@ -21,6 +24,12 @@ public class Projectile : MonoBehaviour {
 
+
+
+        gameObject.transform.position += orignalDirection * speed * Time.timeScale;
```
It's not supposed to be a dumb projectile, it's supposed to track *slightly* rather than perfectly.

linxcat reviewed (approved):
Still some issues with GetComponent when you DON'T need it. Know your reasons for using a Get, it's not a free operation.

Commit:
Merge branch 'prod' into absorbReflect
